### ISSUE TITLE
How do I create an application password using the graph api 2.0?

### ISSUE THREAD

**Issue by carlosscastro** [microsoftgraph/microsoft-graph-docs#3579]
From the fields above, I don't know what to enter in customKeyIdentifier. I also can't find any example on how to create an application with a password. I tried using POST https://graph.microsoft.com/beta/applications. The body I'm using is this:

```
{
  "displayName": "My App"
  "passwordCredentials": [
    {
        "endDateTime": "2020-01-01T00:00:00Z",
        "keyId": "5ac1014f-bfdd-49db-8a7c-e4bc48658e03",
        "startDateTime": "2014-01-01T00:00:00Z",
        "secretText": "5ac1014f-bfdd-49db-8a7c-e4bc48658e02",
        "hint": "dffd"
    }
  ]
}
```

Could you please provide me a way to create an application with a password?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 5d7be764-04c0-883e-25e2-0ee2966b79db
* Version Independent ID: 4adbac8a-6133-3c55-d25d-1134d770ee11
* Content: [passwordCredential resource type - Microsoft Graph beta](https://docs.microsoft.com/en-us/graph/api/resources/passwordcredential?view=graph-rest-beta)
* Content Source: [api-reference/beta/resources/passwordcredential.md](https://github.com/microsoftgraph/microsoft-graph-docs/blob/master/api-reference/beta/resources/passwordcredential.md)
* GitHub Login: @jthake-msft
* Microsoft Alias: **jthake-msft**

**Comment by jthake-msft** [comment]
@sureshja is there any chance you can take a look at this please?

**Comment by carlosscastro** [comment]
I found this and I'll try it now; https://stackoverflow.com/questions/51353442/how-to-create-msa-id-password-for-creating-azure-bot-service-using-sdk/51359149#51359149

However I think that it is pretty hard to understand what each field is supposed to be from the documentation, which is still an issue.

**Comment by jthake-msft** [comment]
@lleonard-msft  ?

**Comment by carlosscastro** [comment]
FYI, the stack overflow didn't work for me. Here is an example request for POST beta/applications

```
{
  "allowPublicClient": true,
  "displayName": "ccastrotestgraphapi",
  passwordCredentials:
  [{
  "customKeyIdentifier": null,
  "endDateTime": "2019-07-01T00:00:00Z",
  "keyId": "b5b2920e-a47c-43b7-91ef-25ae96fddddd", 
  "startDateTime": "2018-07-01T00:00:00Z",
  "secretText": "123QWER!@#$qwweerzcxv2",
  "hint": "string"
}]
}
```

and I get an error that does not help me much as I don't know which piece of my request is wrong, and the documentation does not state the range or meaning of these fields.

```

{
    "error": {
        "code": "Request_BadRequest",
        "message": "One or more property values specified are invalid.",
        "innerError": {
            "request-id": "390574a1-fa22-4586-b73c-87aa051ec0b9",
            "date": "2018-12-10T18:31:22"
        }
    }
}
```

**Comment by suddenelfilio** [comment]
you should use the PATCH verb and the also add the line

 "@odata.context": "https://graph.microsoft.com/beta/$metadata#applications/$entity", to your json object in the request body. 

the URL you'll send the request to must be https://graph.microsoft.com/beta/applications/**{id}**
make sure to replace the {id} part with the id of your application. The id not the appId! You can get it by listing all applications via the graph.

Small note: a single application can only have a maximum of 3 application passwords it seems as trying to add more will result in an error

**Comment by acds** [comment]
Do if you have 3 password credentials how do you patch to remove or replace one ? Use an existing KeyID? Where is the “3 application passwords fact documented”?

**Comment by suddenelfilio** [comment]
Well I tried adding more than 3 password and got an error from the graph api

**Comment by acds** [comment]
Good to know, if guess one can validate vs what’s possible via the portal as behaviors should align?

Any idea how to remove or replace a passwordCredential?

**Comment by carlosscastro** [comment]
@suddenelfilio Thanks for your response, I just came back to this.

You mentioned above "you should use the PATCH verb and the also add the line". However, it sounds  inefficient to make 2 network calls to operate on the same object. Is this because it was designed to work this way, or because the POST does not work correctly?

Thanks!

@stevengum fyi

**Comment by acds** [comment]
As an aside, you can create 100’s of password credentials - I know as I’ve done it.

You need to supply null as the keyId or you will get a bad request.

If you do a GET after the PATCH you can diff and get the keyId of you new key credential.


If you PATCH and keyId is not null the secret value must be null and vice versa.

I don’t  think the API supports POST at all.

**Comment by JoeBrockhaus** [comment]
@acds 

> If you PATCH and keyId is not null the secret value must be null and vice versa.

Is this to say if `keyId` is _**not null**_, it will ignore changes to that Password?

In other words: To add a new Password, first list them, add a new item to the returned array, then PATCH the array back.

**Comment by acds** [comment]
@JoelBrochaus yes you can’t change the secret of an existing key credential (value not null, keyId is null) it seems.

You can add a new one, but to retain the old one(s) you need to supply the prior keyId (not null) and value (null).

Essentially one can use this to rotate keys.  I wrote an function to:
1. Get Credentials for application (my JSON classes just cover KeyCredentials structure)
1. Add a new keyCredential (Patch above with new key credential added to the list)
1. Get Credentials.
1. Diff the two lists of credentials. Ordinarily there will be the difference of the new credential added. From which you can retrieve it’s Id. (If this was no so I’d start again on the basis one has something else manipulating the keys - possible manual intervention)
1. It then gets a current KeyVault key and identifies the old KeyId from its tag (see below)
1. Update the KeyVault key with the new credential value, and tag with the new keyId.
1. On a successful KeyVault update then Get the Key Credentials again and re-diff with the origin set to ensure nothing else changed.
1. Identify the old keyId from the prior KeyValut secret identified above, and remove it from the list, or just leave the new KeyID.
1. Re Patch the key credentials to remove the old key(s).

This is deployed in a continuous web job rotating the keys every hour. Client code pulls the key from the KeyVault (with known permissioned credentials) to authenticate against the application.

Hope that is useful. Can’t share code as it is client work product.

Also @carlosscastro per original question, I was able to set/get the customKeyIdentifier by converting a string in to an ASCII byte array, and base64 enciding it, and in reverse de-baee64 encoding and reconstruct the string from the ASCII byte array...

**Comment by JoeBrockhaus** [comment]
I wonder if there is a functional difference with `keyCredentials` vs `passwordCredentials`. Also worth noting that I'm using the `azure-graph` npm package, so there may be some differences in endpoints/behavior because of that as well, since I believe that operates against the v1 api. 

These payloads resulted in 400s for me:
-  `[ { keyId: '1234' }, { endDate: 'a-valid-date', value: 'password' } ]`
-  `[ { keyId: '1234', value: null }, { endDate: 'a-valid-date', value: 'password' } ]`

Whereas this payload succeeds, albeit deleting previous `passwordCredentials`:
-  `[ { endDate: 'a-valid-date', value: 'password' } ]`

Reviewing the [Application API documentation](https://docs.microsoft.com/en-us/previous-versions/azure/ad/graph/api/entity-and-complex-type-reference#application-entity), it appears a PATCH is supported, however it appears in practice it is an overwrite, not partial update. The `passwordCredentials` Declared Property indicates support for `POST`, `GET`, `PATCH`

**Comment by acds** [comment]
Yes I’ve been using the Beta API as recommended by a MS Azure consultant...

Albeit it won’t be beta at some point and endpoints will need to change.

**Comment by Hong-Revo** [comment]
@JoeBrockhaus you're using the old Active Directory Graph API (host is https://graph.windows.net), not MS Graph (host is https://graph.microsoft.com).

I'm also unable to create a password via the REST API, whether in the initial POST or a later PATCH. The passwordCredentials field I'm passing is

```
{
    "customKeyIdentifier": "awBlAHkAMQA=",  // base64/UTF16-LE encoded "key1"
    "endDateTime": "2020-03-06T22:18:32Z",
    "secretText": "secret"
}
```

The error I get each time is `Property passwordCredentials in payload has a value that does not match schema.`

**Comment by Hong-Revo** [comment]
Nevermind, I figured it out. I was sending a single password credential object, but passwordCredentials expects an array of such objects. The API seems to be working correctly.

**Comment by acds** [comment]
FYI - the PATCH behavior deems to have changed. We get a 400 - Bad Request response the t has worked fine for over 10 months!

```JSON
{
	"passwordCredentials":
		[
			{
				"customKeyIdentifier":"base64/UTF16-encoded-name",
				"endDateTime": "2019-11-02T17:32:09.9638578+00:00",
				"keyId": null,
				"startDateTime": null,
				"secretText": "some-secret-text",
				"hint": null,
				"displayName":"NewKey-01-Nov-2019 10:32:09"
			}
		]
}
```

The expectation is that the `keyId` will be set once it's created, and that until now `startDateTime` did not need to be specified, and neither did `hint` - such that existing hints are maintained from a prior request of existing keys.
 
Anyone else seeing issues ?

**Comment by Hong-Revo** [comment]
Yep, also seeing an error trying to create a new app:

```
Bad request (400).
New password credentials must be generated with service actions.
```

**Comment by acds** [comment]
Hi @Hong-Revo where do you get the more specific error - I looked in the HTTP Response and don't see anything similar - `New password credentials must be generated with service actions`

What are `service actions`?

**Comment by sureshja** [comment]
In order to increase the security profile of the applications, MSGraph blocks developers setting custom passwords. The addPassword method (a.k.a service actions) generate strong passwords for them. 

@carlosscastro, Have you looked at the example from this doc reference? https://docs.microsoft.com/en-us/graph/api/application-addpassword?view=graph-rest-beta&tabs=http

You need not provide a POST body unless you want to have a friendly displayName set to a password for identification later.

**Comment by Hong-Revo** [comment]
@sureshja Thanks. It looks like the app/service principal API is now in the 1.0 API as well, are there any bits still in beta that haven't been finalised?

**Comment by Hong-Revo** [comment]
Also... that call creates a new password, but how do you update an existing one (eg if it's about to expire)? I tried calling `addPassword` with the keyId of an existing password, but it created a new one instead. Are we just supposed to delete old passwords as they expire?

**Comment by acds** [comment]
Hi @sureshja where is a good place to find release notes of the graph api changes. 

Microsoft documentation is a bit of a maze and having the ability to subscribe to change notification would be very valuable as this change has caught us off guard.

Thanks!

**Comment by acds** [comment]
@sureshja how is one supposed to set the `PaswordCredential:CustomKeyIdentifier`?

As a PATCH after the `addPassword`?

**Comment by sureshja** [comment]
@Hong-Revo - Applications is in v1.0. ServicePrincipal is in the final stages of development before I get to v1.0. As we block setting or updating passwords using POST or PATCH, developers should take advantage of [addPassword ](https://docs.microsoft.com/en-us/graph/api/application-addpassword?view=graph-rest-1.0&tabs=http) and [removePassword ](https://docs.microsoft.com/en-us/graph/api/application-removepassword?view=graph-rest-1.0&tabs=http) to rotate secrets.

**Comment by sureshja** [comment]
@acds - `customKeyIdentifier` is deprecated. If you want to set a friendly name to the secrets you are adding, please use `displayName` instead. If you are setting secrets as part of a POST operation at app creation, you will want to separate the secret creation in a subsequent POST call to app creation using [addPassword](https://docs.microsoft.com/en-us/graph/api/application-addpassword?view=graph-rest-1.0&tabs=http).

**Comment by Hong-Revo** [comment]
Thanks @sureshja, do you know roughly when servicePrincipal will be in 1.0?

**Comment by davidmu1** [comment]
It seems as if this issue has been resolved. the add and remove password APIs are now available in both beta and v1.0. No change identified for the page. Closing the issue.
